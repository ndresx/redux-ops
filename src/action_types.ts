export const prefix = '@@redux-ops/';

export const START = `${prefix}START`;
export const UPDATE = `${prefix}UPDATE`;
export const DELETE = `${prefix}DELETE`;
export const CLEAR = `${prefix}CLEAR`;

export const BLUEPRINT = `${prefix}BLUEPRINT`;
