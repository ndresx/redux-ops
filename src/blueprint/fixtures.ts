import { AnyAction } from 'redux';
import { OpBlueprint, OpBlueprintFn } from './typedefs';

export interface MovieFetcherOp extends OpBlueprint {
  readonly start: OpBlueprintFn<typeof fetchMovies>;
  readonly success: OpBlueprintFn<typeof didFetchMovies>;
}

export const movies = [
  { id: 1, name: 'The Dark Knight' },
  { id: 2, name: 'Jurassic World' },
  { id: 3, name: 'Avatar' },
];

export const FETCH_MOVIES = 'FETCH_MOVIES';

export function fetchMovies(genre: string, page?: number): AnyAction {
  return {
    type: FETCH_MOVIES,
    payload: { genre, page },
  };
}

export function didFetchMovies(movies: object): AnyAction {
  return {
    type: `${FETCH_MOVIES}_SUCCESS_CUSTOM`,
    payload: { movies },
  };
}
