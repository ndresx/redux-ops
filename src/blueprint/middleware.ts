import { Middleware } from 'redux';

import * as actionTypes from '../action_types';
import { OpStatus, Operation } from '../typedefs';
import { OpsBlueprintActionData, OpsBlueprintOriginalAction } from './typedefs';

function createOpsMiddleware(): Middleware {
  return ({ dispatch }) => next => action => {
    if (action[actionTypes.prefix] && action[actionTypes.prefix].operationAction) {
      const blueprintAction: OpsBlueprintActionData = action[actionTypes.prefix];
      const { operationAction, originalAction, broadcast, uniqueId } = blueprintAction;
      const opStatus = operationAction.payload.status;
      let operationActionMod = { ...operationAction };
      let broadcastAction: OpsBlueprintOriginalAction | undefined = undefined;

      if (broadcast && !originalAction) {
        let broadcastActionType = operationActionMod.type;

        if (opStatus && broadcastActionType === actionTypes.UPDATE) {
          broadcastActionType = opStatus.toUpperCase();
        } else {
          broadcastActionType = broadcastActionType.replace(actionTypes.prefix, '');
        }

        broadcastAction = {
          type: `${operationActionMod.payload.id}_${broadcastActionType}`,
          payload: { ...operationActionMod.payload },
        };

        if (opStatus !== OpStatus.Delete) {
          // Create a custom action based on the operation's status
          operationActionMod = {
            ...operationActionMod,
            payload: {
              ...operationActionMod.payload,
              data: undefined,
            } as Operation<undefined>,
          };
        }
      }

      // Exchange original op action id with unique id
      if (uniqueId) {
        operationActionMod = {
          ...operationActionMod,
          payload: {
            ...operationActionMod.payload,
            id: uniqueId,
          },
        };

        if (broadcastAction) {
          broadcastAction[actionTypes.prefix] = { uniqueId };
        }
      }

      switch (opStatus) {
        case OpStatus.Start: {
          dispatch(operationActionMod);
          originalAction && dispatch(originalAction);
          broadcastAction && dispatch(broadcastAction);
          break;
        }
        default: {
          originalAction && dispatch(originalAction);
          broadcastAction && dispatch(broadcastAction);
          dispatch(operationActionMod);
        }
      }
    }

    return next(action);
  };
}

export default createOpsMiddleware();
