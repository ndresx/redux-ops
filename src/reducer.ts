import { Reducer } from 'redux';

import * as actionTypes from './action_types';
import { OpsReducerHandler, OpsState, OperationAction, ResetOperationsAction } from './typedefs';

export const defaultState: OpsState = {};

const updateOperation: OpsReducerHandler<OperationAction> = (state, action) => {
  const { payload } = action;
  return { ...state, [payload.id]: { ...payload } };
};

const deleteOperation: OpsReducerHandler<OperationAction> = (state, action) => {
  const newState = { ...state };
  delete newState[action.payload.id];
  return newState;
};

const resetOperations: OpsReducerHandler<ResetOperationsAction> = () => ({ ...defaultState });

const handlers = {
  [actionTypes.START]: updateOperation,
  [actionTypes.UPDATE]: updateOperation,
  [actionTypes.DELETE]: deleteOperation,
  [actionTypes.RESET]: resetOperations,
};

const reducer: Reducer = (state = defaultState, action: any) => {
  const handler = handlers[action.type];
  return handler ? handler(state, action) : state;
};

export default reducer;
