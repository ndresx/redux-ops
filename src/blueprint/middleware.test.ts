import { combineReducers, createStore, AnyAction } from 'redux';

import { createBlueprint, opsUnique, opsBroadcast, getUniqueId } from './index';
import { fetchMovies, didFetchMovies, FETCH_MOVIES, movies, MovieFetcherOp } from './fixtures';
import { prefix } from '../action_types';
import reducer from '../reducer';
import middleware from './middleware';

const store = createStore(combineReducers({ ops: reducer }));

describe('middleware', () => {
  let dispatch: jest.Mock;
  let next: jest.Mock;

  beforeEach(() => {
    dispatch = jest.fn();
    next = jest.fn();
  });

  function testMiddleware(action: AnyAction) {
    middleware({ dispatch, getState: store.getState })(next)(action);
  }

  function expectNext(action: AnyAction) {
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith(action);
  }

  it('should pass through random action', () => {
    const action = { type: 'RANDOM' };

    testMiddleware(action);
    expect(dispatch).not.toHaveBeenCalled();

    expectNext(action);
  });

  it('should process blueprint action', () => {
    const blueprint = createBlueprint(FETCH_MOVIES);
    const action = blueprint.start();

    testMiddleware(action);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch.mock.calls).toEqual([[action[prefix].operationAction]]);

    expectNext(action);
  });

  it('should process blueprint start action with always preceding op action', () => {
    const blueprint = createBlueprint(FETCH_MOVIES, { start: fetchMovies });
    const action = blueprint.start();

    testMiddleware(action);
    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(dispatch.mock.calls).toEqual([
      [action[prefix].operationAction],
      [action[prefix].originalAction],
    ]);

    expectNext(action);
  });

  it('should process blueprint custom action(s) with subsequent op action', () => {
    const blueprint = createBlueprint(FETCH_MOVIES, { success: didFetchMovies });
    const action = blueprint.success();

    testMiddleware(action);
    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(dispatch.mock.calls).toEqual([
      [action[prefix].originalAction],
      [action[prefix].operationAction],
    ]);

    expectNext(action);
  });

  it('should process blueprint delete action', () => {
    const blueprint = createBlueprint(FETCH_MOVIES);
    const action = blueprint.delete();

    testMiddleware(action);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(action[prefix].operationAction);

    expectNext(action);
  });

  describe('unique id', () => {
    it('should overwrite original op action id with unique id', () => {
      const blueprint = createBlueprint(FETCH_MOVIES);
      const action = opsUnique(blueprint.start());

      testMiddleware(action);
      expect(dispatch.mock.calls).toMatchSnapshot();

      const dispatchActionId = dispatch.mock.calls[0][0].payload.id;
      expect(dispatchActionId).not.toBe(action[prefix].operationAction.payload.id);
      expect(dispatchActionId).toMatchInlineSnapshot(`"@@redux-ops/FETCH_MOVIES_1"`);

      expectNext(action);
    });

    it('should attach unique id to custom action', () => {
      const blueprint = createBlueprint<MovieFetcherOp>(FETCH_MOVIES, {
        start: fetchMovies,
      });
      const action = opsUnique(blueprint.start('Science-Fiction'));

      testMiddleware(action);
      expect(dispatch.mock.calls).toMatchSnapshot();

      const customAction = dispatch.mock.calls[1][0];
      expect(customAction[prefix]).toEqual({ uniqueId: getUniqueId(action) });

      expectNext(action);
    });

    it('should attach unique id to broadcast action', () => {
      const blueprint = createBlueprint<MovieFetcherOp>(FETCH_MOVIES, {
        start: fetchMovies,
      });
      const action = opsUnique(opsBroadcast(blueprint.start('Science-Fiction')));

      testMiddleware(action);
      expect(dispatch.mock.calls).toMatchSnapshot();

      const broadcastAction = dispatch.mock.calls[1][0];
      expect(broadcastAction[prefix]).toEqual({ uniqueId: getUniqueId(action) });

      expectNext(action);
    });
  });

  describe('broadcast', () => {
    it('should broadcast blueprint action by using the default action creator', () => {
      const blueprint = createBlueprint(FETCH_MOVIES);
      const action = opsBroadcast(blueprint.start());

      testMiddleware(action);
      expect(dispatch.mock.calls).toMatchSnapshot();

      expectNext(action);
    });

    it('should broadcast unique blueprint action by using the default action creator', () => {
      const blueprint = createBlueprint(FETCH_MOVIES);
      const action = opsUnique(opsBroadcast(blueprint.start()));

      testMiddleware(action);
      expect(dispatch.mock.calls).toMatchSnapshot();

      expectNext(action);
    });

    it('should broadcast an updating blueprint action by using the default action creator', () => {
      const blueprint = createBlueprint<MovieFetcherOp>(FETCH_MOVIES);
      const action = opsUnique(opsBroadcast(blueprint.success(movies)));

      testMiddleware(action);
      expect(dispatch.mock.calls).toMatchSnapshot();

      expectNext(action);
    });

    it('should broadcast blueprint action by using a custom action creator', () => {
      const blueprint = createBlueprint<MovieFetcherOp>(FETCH_MOVIES, { success: didFetchMovies });
      const action = opsBroadcast(blueprint.success(movies));

      testMiddleware(action);
      expect(dispatch.mock.calls).toMatchSnapshot();

      expectNext(action);
    });

    it('should broadcast unique blueprint action by using a custom action creator', () => {
      const blueprint = createBlueprint<MovieFetcherOp>(FETCH_MOVIES, { success: didFetchMovies });
      const action = opsUnique(opsBroadcast(blueprint.success(movies)));

      testMiddleware(action);
      expect(dispatch.mock.calls).toMatchSnapshot();

      expectNext(action);
    });

    it('should broadcast delete action', () => {
      const blueprint = createBlueprint(FETCH_MOVIES);
      const action = opsBroadcast(blueprint.delete());

      testMiddleware(action);
      expect(dispatch.mock.calls).toMatchSnapshot();

      expectNext(action);
    });
  });
});
