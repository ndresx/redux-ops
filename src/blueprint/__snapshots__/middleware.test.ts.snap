// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware broadcast should broadcast an updating blueprint action by using the default action creator 1`] = `
Array [
  Array [
    Object {
      "@@redux-ops/": Object {
        "uniqueId": "@@redux-ops/FETCH_MOVIES_5",
      },
      "payload": Object {
        "data": Array [
          Object {
            "id": 1,
            "name": "Jurassic World",
          },
          Object {
            "id": 2,
            "name": "The Dark Knight",
          },
          Object {
            "id": 3,
            "name": "Avatar",
          },
        ],
        "id": "FETCH_MOVIES",
        "status": "success",
      },
      "type": "FETCH_MOVIES_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "@@redux-ops/FETCH_MOVIES_5",
        "status": "success",
      },
      "type": "@@redux-ops/UPDATE",
    },
  ],
]
`;

exports[`middleware broadcast should broadcast blueprint action by using a custom action creator 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "movies": Array [
          Object {
            "id": 1,
            "name": "Jurassic World",
          },
          Object {
            "id": 2,
            "name": "The Dark Knight",
          },
          Object {
            "id": 3,
            "name": "Avatar",
          },
        ],
      },
      "type": "FETCH_MOVIES_SUCCESS_CUSTOM",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "success",
      },
      "type": "@@redux-ops/UPDATE",
    },
  ],
]
`;

exports[`middleware broadcast should broadcast blueprint action by using the default action creator 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "start",
      },
      "type": "@@redux-ops/START",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "start",
      },
      "type": "FETCH_MOVIES_START",
    },
  ],
]
`;

exports[`middleware broadcast should broadcast delete action 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "delete",
      },
      "type": "FETCH_MOVIES_DELETE",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "delete",
      },
      "type": "@@redux-ops/DELETE",
    },
  ],
]
`;

exports[`middleware broadcast should broadcast unique blueprint action by using a custom action creator 1`] = `
Array [
  Array [
    Object {
      "@@redux-ops/": Object {
        "uniqueId": "@@redux-ops/FETCH_MOVIES_6",
      },
      "payload": Object {
        "movies": Array [
          Object {
            "id": 1,
            "name": "Jurassic World",
          },
          Object {
            "id": 2,
            "name": "The Dark Knight",
          },
          Object {
            "id": 3,
            "name": "Avatar",
          },
        ],
      },
      "type": "FETCH_MOVIES_SUCCESS_CUSTOM",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "@@redux-ops/FETCH_MOVIES_6",
        "status": "success",
      },
      "type": "@@redux-ops/UPDATE",
    },
  ],
]
`;

exports[`middleware broadcast should broadcast unique blueprint action by using the default action creator 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "@@redux-ops/FETCH_MOVIES_4",
        "status": "start",
      },
      "type": "@@redux-ops/START",
    },
  ],
  Array [
    Object {
      "@@redux-ops/": Object {
        "uniqueId": "@@redux-ops/FETCH_MOVIES_4",
      },
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "start",
      },
      "type": "FETCH_MOVIES_START",
    },
  ],
]
`;

exports[`middleware unique id should attach unique id to broadcast action 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "@@redux-ops/FETCH_MOVIES_3",
        "status": "start",
      },
      "type": "@@redux-ops/START",
    },
  ],
  Array [
    Object {
      "@@redux-ops/": Object {
        "uniqueId": "@@redux-ops/FETCH_MOVIES_3",
      },
      "payload": Object {
        "genre": "Science-Fiction",
        "page": undefined,
      },
      "type": "FETCH_MOVIES",
    },
  ],
]
`;

exports[`middleware unique id should attach unique id to custom action 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "@@redux-ops/FETCH_MOVIES_2",
        "status": "start",
      },
      "type": "@@redux-ops/START",
    },
  ],
  Array [
    Object {
      "@@redux-ops/": Object {
        "uniqueId": "@@redux-ops/FETCH_MOVIES_2",
      },
      "payload": Object {
        "genre": "Science-Fiction",
        "page": undefined,
      },
      "type": "FETCH_MOVIES",
    },
  ],
]
`;

exports[`middleware unique id should overwrite original op action id with unique id 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "data": undefined,
        "id": "@@redux-ops/FETCH_MOVIES_1",
        "status": "start",
      },
      "type": "@@redux-ops/START",
    },
  ],
]
`;
